File: libgraffiti/src/dom/Node.zig
const std = @import("std");
const Document = @import("dom.zig").Document;
const NodeId = @import("dom.zig").NodeId;
const NodeType = @import("dom.zig").NodeType;
const Element = @import("dom.zig").Element;
const Text = @import("dom.zig").Text;
const Comment = @import("dom.zig").Comment;

pub const NodeId = usize;

pub const NodeType = enum(u32) {
    element = 1,
    text = 3,
    comment = 8,
    document = 9,
    document_fragment = 11,
};

pub const Node = struct {
    document: *Document,
    id: NodeId, // TODO: or lookup-table in Document?

    parent_node: ?*Node = null,
    first_child: ?*Node = null,
    next_sibling: ?*Node = null,

    data: union(NodeType) {
        element: *Element,
        text: *Text,
        comment: *Comment,
        document: *Document,
        document_fragment,
    },

    const Self = @This();

    pub fn nodeType(self: *Self) NodeType {
        return self.data;
    }

    pub fn text(self: *Self) ?*Text {
        return switch (self.data) {
            .text => |ptr| ptr,
            else => null,
        };
    }

    pub fn element(self: *Self) ?*Element {
        return switch (self.data) {
            .element => |ptr| ptr,
            else => null,
        };
    }

    pub fn childNodes(self: *Self) ChildNodesIter {
        return .{ .next = self.first_child };
    }

    pub fn appendChild(self: *Self, child: *Self) void {
        // TODO: assert self.data != .text and child.parent == null

        if (self.first_child) |first| {
            var last = first;
            while (last.next_sibling) |n| last = n;

            last.next_sibling = child;
        } else {
            self.first_child = child;
        }

        child.parent_node = self;
    }

    // TODO: insertBefore()
    // TODO: replaceChild()
    // TODO: removeChild()
    // TODO: querySelector()
    // TODO: querySelectorAll()
};

const ChildNodesIter = struct {
    next: ?*Node,

    pub fn next(self: ChildNodesIter) ?*Node {
        defer self.next = next.next_sibling;
        return self.next;
    }
};

File: lib/core/Node.js
// TODO: https://github.com/cztomsik/graffiti/blob/bd1dfe61d3d7b5bfbf9184ecfb9e068dda982a60/src/dom/Node.ts

import { native } from './native.js'
import { EventTarget } from './EventTarget.js'

export class Node extends EventTarget {
  // TODO: NodeList
  get childNodes() {
    const res = []
    for (let n = this.firstChild; n; n = n.nextSibling) res.push(n)
    return res
  }

  appendChild(child) {
    native.Node_appendChild(this, child)

    if (this.lastChild) {
      this.lastChild.nextSibling = child
      child.previousSibling = this.lastChild
    } else {
      this.firstChild = child
    }

    this.lastChild = child
    child.parentNode = this

    return child
  }

  insertBefore(child, before) {
    if (!before) {
      return this.appendChild(child)
    }

    if (before.previousSibling) {
      before.previousSibling.nextSibling = child
      child.previousSibling = before.previousSibling
    } else {
      this.firstChild = child
    }

    child.nextSibling = before
    before.previousSibling = child
    child.parentNode = this

    native.Node_insertBefore(this, child, before)
    return child
  }

  replaceChild(child, oldChild) {
    this.insertBefore(child, oldChild)
    return this.removeChild(oldChild)
  }

  removeChild(child) {
    native.Node_removeChild(this, child)

    if (child.previousSibling) {
      child.previousSibling.nextSibling = child.nextSibling
    } else {
      this.firstChild = child.nextSibling
    }

    if (child.nextSibling) {
      child.nextSibling.previousSibling = child.previousSibling
    } else {
      this.lastChild = child.previousSibling
    }

    child.nextSibling = null
    child.previousSibling = null
    child.parentNode = null

    return child
  }

  hasChildNodes() {
    return this.childNodes.length > 0
  }

  get parentElement() {
    return this.parentNode instanceof HTMLElement ? this.parentNode : null
  }

  // overridden by CharacterData
  // comment.textContent should return a value but it
  // shouldn't be part of element.textContent
  get textContent() {
    return this.childNodes
      .filter(c => c.nodeType == Node.ELEMENT_NODE || c.nodeType == Node.TEXT_NODE)
      .map(c => c.textContent)
      .join('')
  }

  // overridden by CharacterData
  set textContent(v) {
    this.childNodes.forEach(c => c.remove())

    // note we can't just update already present text node because it has to remain untouched
    this.appendChild(this.ownerDocument.createTextNode('' + v))
  }

  get isConnected() {
    return this.parentNode?.isConnected ?? false
  }

  contains(other) {
    while (other) {
      if (other === this) {
        return true
      }

      other = other.parentNode
    }

    return false
  }

  // ---
  // ParentNode:

  get children() {
    // TODO: HTMLCollection
    return this.childNodes.filter(c => c.nodeType === Node.ELEMENT_NODE)
  }

  get childElementCount() {
    return this.children.length
  }

  get firstElementChild() {
    return this.children[0] ?? null
  }

  get lastElementChild() {
    const { children } = this
    return children[children.length - 1] ?? null
  }

  getElementById(id) {
    return this.querySelector(`#${id}`)
  }

  querySelector(selector) {
    return native.Node_querySelector(this, selector)
  }

  querySelectorAll(selector) {
    return native.Node_querySelectorAll(this, selector)
  }

  getElementsByTagName(tagName) {
    return this.querySelectorAll(tagName)
  }

  getElementsByClassName(className) {
    return this.querySelectorAll(`.${className}`)
  }

  // ---
  // NonDocumentTypeChildNode:

  get nextElementSibling() {
    return (
      this.nextSibling &&
      (this.nextSibling.nodeType === Node.ELEMENT_NODE ? this.nextSibling : this.nextSibling.nextElementSibling)
    )
  }

  get previousElementSibling() {
    return (
      this.previousSibling &&
      (this.previousSibling.nodeType === Node.ELEMENT_NODE
        ? this.previousSibling
        : this.previousSibling.previousElementSibling)
    )
  }

  // node types
  static ELEMENT_NODE = 1
  static TEXT_NODE = 3
  static COMMENT_NODE = 8
  static DOCUMENT_NODE = 9
  static DOCUMENT_FRAGMENT_NODE = 11
}

Node.prototype.parentNode = null
Node.prototype.firstChild = null
Node.prototype.lastChild = null
Node.prototype.previousSibling = null
Node.prototype.nextSibling = null
